# Delivery Time Prediction - Phase 2: Dataset Understanding

## Executive Overview
Welcome to **Phase 2: Dataset Understanding** of the Delivery Time Prediction machine learning project.

The ultimate objective of this project is to build an end-to-end regression model to predict food delivery durations in minutes. Before jumping into data preprocessing, feature engineering, or modeling, it is essential to systematically inspect and understand the raw data schema, feature types, distributions, anomalies, and missing values.

### Objectives of this Phase
1. Load the dataset using Pandas without modifying the raw data.
2. Inspect dimensions, column names, data types, and sample records.
3. Formulate the comprehensive Data Dictionary.
4. Classify columns into Target (Y), Numerical Features, Categorical Features, and Metadata/Identifiers.
5. Perform descriptive statistical profiling.
6. Audit data quality: check for missing values, duplicates, and range anomalies.
7. Synthesize key findings and document recommendations for future phases.

In [1]:
import os
import pandas as pd
import numpy as np

# Configure pandas display settings
pd.set_option('display.max_columns', 20)
pd.set_option('display.width', 1000)

# Resolve dataset path (checks delivery_data.csv with fallback to Food_Delivery_Times.csv)
DATA_PATH = os.path.join('..', 'data', 'delivery_data.csv')
if not os.path.exists(DATA_PATH):
    DATA_PATH = os.path.join('..', 'data', 'Food_Delivery_Times.csv')

print(f"Loading dataset from: {DATA_PATH}")
df = pd.read_csv(DATA_PATH)
print(f"Successfully loaded dataset with {len(df)} records.")

Loading dataset from: ..\data\delivery_data.csv
Successfully loaded dataset with 1000 records.


### Analysis & Observations
* The dataset was successfully loaded directly from the `data/` directory.
* Pandas loaded the raw CSV without any in-place mutation, preserving original data integrity as required by Phase 2 rules.

## 1. High-Level Dataset Inspection
We inspect the dataset dimensions, schema, and column data types.

In [11]:
# Display dimensions
print(f"Dataset Dimensions: {df.shape[0]} rows x {df.shape[1]} columns\n")

# Display schema information and memory footprint
df.info()

Dataset Dimensions: 1000 rows x 9 columns

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Order_ID                1000 non-null   int64  
 1   Distance_km             1000 non-null   float64
 2   Weather                 970 non-null    str    
 3   Traffic_Level           970 non-null    str    
 4   Time_of_Day             970 non-null    str    
 5   Vehicle_Type            1000 non-null   str    
 6   Preparation_Time_min    1000 non-null   int64  
 7   Courier_Experience_yrs  970 non-null    float64
 8   Delivery_Time_min       1000 non-null   int64  
dtypes: float64(2), int64(3), str(4)
memory usage: 70.4 KB


### Analysis & Observations
* **Dimensions**: The dataset contains exactly **1,000 rows** and **9 columns**.
* **Memory Usage**: The memory footprint is approximately **70.4 KB**, which is lightweight and easily manageable in-memory.
* **Data Types**:
  * 3 integer columns (`int64`): `Order_ID`, `Preparation_Time_min`, `Delivery_Time_min`
  * 2 floating-point columns (`float64`): `Distance_km`, `Courier_Experience_yrs`
  * 4 categorical / string columns (`object` / `str`): `Weather`, `Traffic_Level`, `Time_of_Day`, `Vehicle_Type`
* **Non-Null Counts**: While several columns have 1,000 non-null values, `Weather`, `Traffic_Level`, `Time_of_Day`, and `Courier_Experience_yrs` show 970 non-null values, indicating the presence of missing data.

## 2. Inspecting First and Last Records
Viewing the first and last records confirms the structure and alignment of the data.

In [12]:
# Display first 5 records
print("--- First 5 Records ---")
df.head(5)

--- First 5 Records ---


,Order_ID,Distance_km,Weather,Traffic_Level,Time_of_Day,Vehicle_Type,Preparation_Time_min,Courier_Experience_yrs,Delivery_Time_min
0,522,7.93,Windy,Low,Afternoon,Scooter,12,1.0,43
1,738,16.42,Clear,Medium,Evening,Bike,20,2.0,84
2,741,9.52,Foggy,Low,Night,Scooter,28,1.0,59
3,661,7.44,Rainy,Medium,Afternoon,Scooter,5,1.0,37
4,412,19.03,Clear,Low,Morning,Bike,16,5.0,68


In [14]:
# Display last 5 records
print("--- Last 5 Records ---")
df.tail(5)

--- Last 5 Records ---


,Order_ID,Distance_km,Weather,Traffic_Level,Time_of_Day,Vehicle_Type,Preparation_Time_min,Courier_Experience_yrs,Delivery_Time_min
995,107,8.50,Clear,High,Evening,Car,13,3.0,54
996,271,16.28,Rainy,Low,Morning,Scooter,8,9.0,71
997,861,15.62,Snowy,High,Evening,Scooter,26,2.0,81
998,436,14.17,Clear,Low,Afternoon,Bike,8,0.0,55
999,103,6.63,Foggy,Low,Night,Scooter,24,3.0,58


### Analysis & Observations
* Each record represents a distinct food delivery order.
* Each order contains distance metrics, ambient conditions (weather, traffic, time of day), courier characteristics, and delivery durations.
* Order IDs range from 1 to 1000, representing unique transactional identifiers.

## 3. Data Dictionary & Column Documentation
Below is the formal documentation of each column in the dataset:

| Column Name | Data Type | Role | Business Meaning & Description |
| :--- | :--- | :--- | :--- |
| `Order_ID` | Integer (`int64`) | Metadata / ID | Unique identifier for each order. Arbitrary index; contains no predictive signal. |
| `Distance_km` | Float (`float64`) | Feature ($X$) | Distance in kilometers between the restaurant and the delivery destination. |
| `Weather` | String / Categorical | Feature ($X$) | Weather conditions during delivery (`Clear`, `Rainy`, `Foggy`, `Snowy`, `Windy`). |
| `Traffic_Level` | String / Categorical | Feature ($X$) | Traffic congestion level (`Low`, `Medium`, `High`). |
| `Time_of_Day` | String / Categorical | Feature ($X$) | Period of day the order was placed (`Morning`, `Afternoon`, `Evening`, `Night`). |
| `Vehicle_Type` | String / Categorical | Feature ($X$) | Vehicle used by courier (`Bike`, `Scooter`, `Car`). |
| `Preparation_Time_min` | Integer (`int64`) | Feature ($X$) | Kitchen food preparation time in minutes at the restaurant. |
| `Courier_Experience_yrs` | Float (`float64`) | Feature ($X$) | Number of years of experience the delivery courier has. |
| `Delivery_Time_min` | Integer (`int64`) | **Target ($Y$)** | **Target variable**: Total delivery duration in minutes from dispatch to customer. |

## 4. Feature & Target Variable Classification
We programmatically categorize the columns into Target ($Y$), Numerical Features, Categorical Features, and Identifiers.

In [15]:
# Define target variable
target_col = 'Delivery_Time_min'

# Define metadata / identifier column
id_col = 'Order_ID'

# Identify numerical feature columns
numerical_features = [
    col for col in df.select_dtypes(include=[np.number]).columns
    if col not in [target_col, id_col]
]

# Identify categorical feature columns
categorical_features = df.select_dtypes(include=['object', 'str']).columns.tolist()

print(f"Target Variable (Y):\n  -> {target_col}\n")
print(f"Identifier (Excluded from features):\n  -> {id_col}\n")
print(f"Numerical Features ({len(numerical_features)}):\n  -> {numerical_features}\n")
print(f"Categorical Features ({len(categorical_features)}):\n  -> {categorical_features}")

Target Variable (Y):
  -> Delivery_Time_min

Identifier (Excluded from features):
  -> Order_ID

Numerical Features (3):
  -> ['Distance_km', 'Preparation_Time_min', 'Courier_Experience_yrs']

Categorical Features (4):
  -> ['Weather', 'Traffic_Level', 'Time_of_Day', 'Vehicle_Type']


### Analysis & Observations
* **Supervised Learning Task**: This is a **Continuous Regression** problem because the target `Delivery_Time_min` is numeric and continuous.
* **Predictor Set ($X$)**:
  * 3 Numerical inputs: `Distance_km`, `Preparation_Time_min`, `Courier_Experience_yrs`
  * 4 Categorical inputs: `Weather`, `Traffic_Level`, `Time_of_Day`, `Vehicle_Type`
* **Exclusion of `Order_ID`**: `Order_ID` must strictly be excluded from model training to prevent target leakage and spurious index correlations.

## 5. Descriptive Statistical Profiling
We examine the central tendency, dispersion, and quartiles of numerical columns.

In [16]:
# Statistical summary for numerical columns
df[[id_col] + numerical_features + [target_col]].describe().T

,count,mean,std,min,25%,50%,75%,max
Order_ID,1000.0,500.500000,288.819436,1.00,250.750,500.50,750.2500,1000.00
Distance_km,1000.0,10.059970,5.696656,0.59,5.105,10.19,15.0175,19.99
Preparation_Time_min,1000.0,16.982000,7.204553,5.00,11.000,17.00,23.0000,29.00
Courier_Experience_yrs,970.0,4.579381,2.914394,0.00,2.000,5.00,7.0000,9.00
Delivery_Time_min,1000.0,56.732000,22.070915,8.00,41.000,55.50,71.0000,153.00


### Analysis & Observations
* **`Distance_km`**:
  * Ranges from **0.59 km** to **19.99 km** with an average of **10.06 km** and median of **10.19 km**.
  * Shows a uniform spread across standard urban delivery radii.
* **`Preparation_Time_min`**:
  * Ranges from **5.0 minutes** to **29.0 minutes** with an average of **16.98 minutes** and median of **17.0 minutes**.
  * Matches typical fast-food / restaurant kitchen fulfillment times.
* **`Courier_Experience_yrs`**:
  * Ranges from **0.0 years** (novice) to **9.0 years** (highly experienced) with an average of **4.58 years**.
* **Target (`Delivery_Time_min`)**:
  * Minimum delivery time: **8 minutes**.
  * Median delivery time: **55.5 minutes**.
  * Maximum delivery time: **153 minutes** (potential heavy traffic / bad weather orders).
  * Standard deviation: **22.07 minutes**, indicating significant variance suitable for machine learning regression modeling.

## 6. Categorical Feature Inspection & Cardinality
We analyze the unique categories and frequency distribution of each categorical feature.

In [17]:
# Inspect unique values and distribution for each categorical column
for col in categorical_features:
    unique_vals = df[col].unique()
    n_unique = df[col].nunique(dropna=False)
    print(f"=== {col} (Unique values: {n_unique}) ===")
    val_counts = df[col].value_counts(dropna=False)
    pct_dist = df[col].value_counts(dropna=False, normalize=True) * 100
    summary_cat = pd.DataFrame({'Count': val_counts, 'Percentage (%)': pct_dist.round(2)})
    print(summary_cat)
    print()

=== Weather (Unique values: 6) ===
         Count  Percentage (%)
Weather                       
Clear      470            47.0
Rainy      204            20.4
Foggy      103            10.3
Snowy       97             9.7
Windy       96             9.6
NaN         30             3.0

=== Traffic_Level (Unique values: 4) ===
               Count  Percentage (%)
Traffic_Level                       
Medium           390            39.0
Low              383            38.3
High             197            19.7
NaN               30             3.0

=== Time_of_Day (Unique values: 5) ===
             Count  Percentage (%)
Time_of_Day                       
Morning        308            30.8
Evening        293            29.3
Afternoon      284            28.4
Night           85             8.5
NaN             30             3.0

=== Vehicle_Type (Unique values: 3) ===
              Count  Percentage (%)
Vehicle_Type                       
Bike            503            50.3
Scooter         302

### Analysis & Observations
* **`Weather`**: 5 distinct categories (`Clear`, `Rainy`, `Foggy`, `Snowy`, `Windy`) + `NaN` (3.0%). `Clear` is the most prevalent condition (47.0%).
* **`Traffic_Level`**: 3 distinct categories (`Low`, `Medium`, `High`) + `NaN` (3.0%). `Medium` (39.0%) and `Low` (38.3%) are balanced, while `High` accounts for 19.7%.
* **`Time_of_Day`**: 4 distinct periods (`Morning`, `Evening`, `Afternoon`, `Night`) + `NaN` (3.0%). `Night` has the lowest volume (8.5%), while Daytime and Evening orders are evenly distributed.
* **`Vehicle_Type`**: 3 distinct categories (`Bike`: 50.3%, `Scooter`: 30.2%, `Car`: 19.5%) with **0 missing values**.
* Cardinalities are low across all categorical columns, making them suitable for one-hot encoding or ordinal encoding in future phases.

## 7. Data Quality Audit: Missing Values & Duplicates
We check for null entries across all fields and duplicate records.

In [8]:
# Missing values count and percentage
null_counts = df.isnull().sum()
null_percent = (df.isnull().sum() / len(df)) * 100
missing_df = pd.DataFrame({
    'Missing Count': null_counts,
    'Missing Percentage (%)': null_percent.round(2)
})
missing_df[missing_df['Missing Count'] > 0]

,Missing Count,Missing Percentage (%)
Weather,30,3.0
Traffic_Level,30,3.0
Time_of_Day,30,3.0
Courier_Experience_yrs,30,3.0


In [9]:
# Check for duplicate rows
duplicate_count = df.duplicated().sum()
print(f"Total Duplicate Rows: {duplicate_count}")

# Check for duplicate Order_IDs
duplicate_orders = df['Order_ID'].duplicated().sum()
print(f"Duplicate Order_IDs: {duplicate_orders}")

Total Duplicate Rows: 0
Duplicate Order_IDs: 0


### Analysis & Observations
* **Missing Values**:
  * Exactly 4 columns contain missing values: `Weather` (30 nulls, 3.0%), `Traffic_Level` (30 nulls, 3.0%), `Time_of_Day` (30 nulls, 3.0%), and `Courier_Experience_yrs` (30 nulls, 3.0%).
  * All other 5 columns (`Order_ID`, `Distance_km`, `Vehicle_Type`, `Preparation_Time_min`, `Delivery_Time_min`) are 100% complete.
  * In total, 117 rows (11.7% of the dataset) contain at least one missing field.
  * Since missingness is only 3% per column, standard imputation strategies (e.g., median for courier experience, mode/most-frequent or 'Missing' token for categoricals) can be evaluated in Phase 4 (Preprocessing).
* **Duplicates**:
  * **0 duplicate rows** found across the entire dataset.
  * **0 duplicate Order_IDs** found; every order represents a distinct transaction.

## 8. Range Validation & Sanity Checks
We verify that numerical values do not violate physical boundaries or business logic.

In [10]:
# Range and sanity checks
sanity_checks = {
    'Negative Distance': (df['Distance_km'] < 0).sum(),
    'Zero or Negative Prep Time': (df['Preparation_Time_min'] <= 0).sum(),
    'Negative Courier Experience': (df['Courier_Experience_yrs'] < 0).sum(),
    'Negative Delivery Time': (df['Delivery_Time_min'] <= 0).sum()
}

for check, violations in sanity_checks.items():
    status = 'PASSED (No violations)' if violations == 0 else f'FAILED ({violations} violations)'
    print(f"{check:30s}: {status}")

Negative Distance             : PASSED (No violations)
Zero or Negative Prep Time    : PASSED (No violations)
Negative Courier Experience   : PASSED (No violations)
Negative Delivery Time        : PASSED (No violations)


### Analysis & Observations
* **Sanity Checks Passed**:
  * No negative distances or zero prep times.
  * No negative courier experience values.
  * No negative or zero delivery times.
* The data conforms to physical delivery constraints, and no corrupt or nonsensical values were detected.

## 9. Executive Summary & Next Phase Recommendations

### Key Dataset Findings
1. **Dataset Scope**: 1,000 delivery orders with 9 attributes (8 candidate inputs + 1 target).
2. **Target Variable**: `Delivery_Time_min` (continuous numeric value ranging from 8 to 153 minutes, mean = 56.73 minutes, median = 55.50 minutes).
3. **Features Overview**:
   * **3 Numerical Features**: `Distance_km` (0.59 - 19.99 km), `Preparation_Time_min` (5 - 29 min), `Courier_Experience_yrs` (0 - 9 yrs).
   * **4 Categorical Features**: `Weather` (5 categories), `Traffic_Level` (3 categories), `Time_of_Day` (4 categories), `Vehicle_Type` (3 categories).
   * **1 Excluded Metadata**: `Order_ID` (to be dropped during modeling).
4. **Data Integrity**:
   * **0 duplicate rows**.
   * Missing values are present in exactly 4 columns at 3.0% each (`Weather`, `Traffic_Level`, `Time_of_Day`, `Courier_Experience_yrs`).
   * No physically impossible values (no negative times or distances).

### Guidance for Subsequent Phases
* **Phase 3 (Exploratory Data Analysis - EDA)**: Investigate bivariate relationships between delivery duration and factors like weather conditions, traffic levels, and distance.
* **Phase 4 (Preprocessing & Feature Engineering)**: Handle missing values via robust imputation (e.g. median for courier experience and mode/unknown token for categoricals), encode categorical variables, and drop `Order_ID`.
* **Phase 5 (Train/Test Split & Modeling)**: Establish a clean stratified or randomized split before fitting transformations to prevent data leakage.

*Phase 2: Dataset Understanding is concluded. No preprocessing, feature engineering, or model training was performed.*